### PR TITLE
HL-394

### PR DESCRIPTION
**Pull request by deBhal**
[Link to ticket](https://dltx.atlassian.net/browse/HL-394)

Also addresses [HL-83](https://dltx.atlassian.net/browse/HL-83)

This code could use some serious prettying up, but it's 2 huge tickets and they're working, so let's get it running and do some refactoring in a followup PR

[HL-83]: https://dltx.atlassian.net/browse/HL-83?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: contracts/OddsLib.sol [hunk 1]
```diff
@@ -53,4 +53,32 @@ library OddsLib {
         // Return the odds need to generate this adjusted payout with the given wager
         return Math.max(1 * PRECISION, (adjustedPayout * PRECISION) / wager);
     }
+
+    // Assuming that the market previously had targetMargin, then after some runners have been scratched, correct the odds of this runner to make up the margin again
+    function rebaseOddsWithScratch(uint256 odds, uint256 scratchedOdds, uint256 targetMargin) external pure returns (uint256) {
+        uint256 newMargin = targetMargin - PRECISION / scratchedOdds;
```
Comment (bitcoinbrisbane) [inline contracts/OddsLib.sol:59]: Should assert `scratchedOdds` is not equal to 0

File: contracts/OddsLib.sol [hunk 2]
```diff
@@ -53,4 +53,32 @@ library OddsLib {
...
+        return odds * margin / PRECISION;
+    }
+
+    function addMargin(uint256 odds, uint256 margin) public pure returns (uint256) {
+        return odds * PRECISION / margin;
+    }
+
+    // Given an array of odds, return the margin
+    function getMargin(uint256[] calldata odds) public pure returns (uint256) {
+        uint256 total = 0;
```
Comment (bitcoinbrisbane) [inline contracts/OddsLib.sol:77]: Save some gas...

```suggestion
        uint256 total;
```

File: contracts/OddsLib.sol [hunk 3]
```diff
@@ -53,4 +53,32 @@ library OddsLib {
...
+        for (uint256 i = 0; i < odds.length; i++) {
```
Comment (bitcoinbrisbane) [inline contracts/OddsLib.sol:78]: Its better to "cache" this value, as reading the .length property uses more gas.

```suggestion
        uint256 count = odds.length;
        for (uint256 i = 0; i < count; i++) {
```

File: scripts/result.py [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+        "id": str(id),
+        "date": date,
+        "location": location,
+        "race": race
+    }
+    return result
+
+
+def get_oracle():
+    response = requests.get('https://horse.link/api/config')
```
Comment (bitcoinbrisbane) [inline scripts/result.py:64]: Check this link, it might be `https://alpha.horse.link/api/config` now.

File: contracts/Market.sol [hunk 5]
```diff
@@ -344,6 +349,34 @@ contract Market is IMarket, Ownable, ERC721 {
...
 
+	function _applyScratchings(uint64 index) internal virtual returns (uint256) {
+		// Get marketId of bet
+		bytes16 marketId = _bets[index].marketId;
+		// Ask the oracle for scratched runners on this market
+		IOracle.Result memory result = IOracle(_oracle).getResult(marketId);
+
+		// Get all scratchings with a timestamp after this bet
+		// Loop through scratchings
+		uint256 scratchedOdds = 0;
```
Comment (bitcoinbrisbane) [inline contracts/Market.sol:360]: Gas saver...

```suggestion
		uint256 scratchedOdds;
```

File: contracts/Market.sol [hunk 6]
```diff
@@ -344,6 +349,34 @@ contract Market is IMarket, Ownable, ERC721 {
...
+		uint256 betCreated = _bets[index].created;
+		for (uint256 i = 0; i < result.scratched.length; i++) {
```
Comment (bitcoinbrisbane) [inline contracts/Market.sol:362]: Same here with the .length.  Create a temp var to store length so the method costs less gas.

**Review by bitcoinbrisbane — CHANGES_REQUESTED** [review]
- Few solidity things to look at
- ~Check the URLs on the API~
- Check double line spacing on python between functions
- ~Bit of duplicate methods.~

**Comment by deBhal** [comment]
I've refactored the shared .ts code and fixed up the `alpha.horse.link` url in `result.py`

**Comment by deBhal** [comment]
I've fixed the https://alpha.horse.link url in the scripts directory as well as `contracts/Market.sol` and `tests/market.test.ts` because we change them in this PR, but there still are some stale references in the `deployments` directory:
`ag -l '\/\/horse.link'
```
deployments/goerli/DaiVault.json
deployments/goerli/DaiMarket.json
deployments/goerli/UsdtMarket.json
deployments/goerli/UsdtVault.json
```

File: contracts/OddsLib.sol [hunk 7]
```diff
@@ -53,4 +53,32 @@ library OddsLib {
...
+    function changeMargin(uint256 odds, uint256 margin, uint256 targetMargin) public pure returns (uint256) {
+        return addMargin(removeMargin(odds, margin), targetMargin);
+    }
+
+    function removeMargin(uint256 odds, uint256 margin) public pure returns (uint256) {
+        return odds * margin / PRECISION;
+    }
+
+    function addMargin(uint256 odds, uint256 margin) public pure returns (uint256) {
+        return odds * PRECISION / margin;
```
Comment (bitcoinbrisbane) [inline contracts/OddsLib.sol:72]: Check != 0

**Review by bitcoinbrisbane — APPROVED** [review]

File: contracts/Market.sol [hunk 8]
```diff
@@ -344,6 +349,36 @@ contract Market is IMarket, Ownable, ERC721 {
...
+		bytes16 marketId = _bets[index].marketId;
+		// Ask the oracle for scratched runners on this market
+		IOracle.Result memory result = IOracle(_oracle).getResult(marketId);
+
+		// Get all scratchings with a timestamp after this bet
+		// Loop through scratchings
+		uint256 scratchedOdds;
+		uint256 betCreated = _bets[index].created;
+		uint256 scratchedCount = result.scratched.length;
+		uint256 scratchCount = scratchedCount;
```
Comment (bitcoinbrisbane) [inline contracts/Market.sol:363]:
```suggestion
		uint256 scratchCount = result.scratched.length;
```

File: contracts/Market.sol [hunk 9]
```diff
@@ -344,6 +349,36 @@ contract Market is IMarket, Ownable, ERC721 {
...
+		for (uint256 i = 0; i < scratchCount; i++) {
+			// If the timestamp of the scratching is after the bet
+			if (result.scratched[i].timestamp > betCreated) {
+				//Sum the odds
+				scratchedOdds += result.scratched[i].odds;
+			}
+		}
+		// Calculate the odds of the bet
+		uint256 originalOdds = _bets[index].payout / _bets[index].amount;
```
Comment (bitcoinbrisbane) [inline contracts/Market.sol:372]: Checkout amount is not zero

**Review by bitcoinbrisbane — CHANGES_REQUESTED** [review]
As discussed

**Review by bitcoinbrisbane — APPROVED** [review]

**Pull request merged**